### PR TITLE
Shells and Periphery

### PR DESCRIPTION
**Pull request by LordFowl**
Adds Shells. Adds tags. Adds IPC subspecies. Adds HK-47's. Does not add IPC construction.

Beyond the obvious review of shells, the exact stats of the IPC subspecies need review. They were largely an afterthought, and quickly overshadowed by the shell project.

### CODE REVIEW

**Comment by LordFowl** [comment]
Added tag persistency and tags to cargo as per Skull's request.

File: code/modules/client/preference_setup/general/01_basic.dm [hunk 1]
```diff
@@ -142,3 +142,30 @@
 			return TOPIC_REFRESH
 
 	return ..()
+
+/datum/category_item/player_setup_item/general/basic/proc/getMinAge(var/age_min)
```
Comment (skull132) [inline code/modules/client/preference_setup/general/01_basic.dm]: Couldn't you have coded this as a species variable instead? And then simply write the getters as:
```
/datum/category_item/player_setup_item/general/basic/proc/getMinAge()
    var/datum/species/mob_species = all_species[pref.species]
    return mob_species.age_min
```

10x prefereable to a massive switch statement.

File: code/modules/client/preference_setup/general/03_body.dm [hunk 2]
```diff
@@ -514,6 +542,52 @@ var/global/list/valid_bloodtypes = list("A+", "A-", "B+", "B-", "AB+", "AB-", "O
...
+				mob_species.tail_animation = null
+			if("Unathi")
+				shell_type = "Unathi"
+				mob_species.tail = "sogtail"
+				mob_species.tail_animation = 'icons/mob/species/unathi/tail.dmi'
+			if("Tajara")
+				shell_type = "Tajara"
+				mob_species.tail = "tajtail"
+				mob_species.tail_animation = 'icons/mob/species/tajaran/tail.dmi'
+			if("Dionaea")
```
Comment (skull132) [inline code/modules/client/preference_setup/general/03_body.dm]: Dionaea shells...?

File: code/modules/client/preference_setup/general/03_body.dm [hunk 3]
```diff
@@ -581,3 +655,15 @@ var/global/list/valid_bloodtypes = list("A+", "A-", "B+", "B-", "AB+", "AB-", "O
 	dat += "</center></body>"
 
 	user << browse(dat, "window=species;size=700x400")
+
+/datum/category_item/player_setup_item/general/body/proc/reset_limbs()
+
+	for(var/organ in pref.organ_data)
+		pref.organ_data[organ] = null
+	while(null in pref.organ_data)
```
Comment (skull132) [inline code/modules/client/preference_setup/general/03_body.dm]: wat. Did you manage to dig this up from somewhere?

```
pref.organ_data.Cut()
```

File: code/modules/client/preferences.dm [hunk 4]
```diff
@@ -149,23 +149,31 @@ datum/preferences
 		save_character()
 
 /datum/preferences/proc/getMinAge(var/age_min)
-	if(species == "Vaurca Worker" || species == "Vaurca Warrior" || species == "Machine" || species == "Diona")
+
+	var/datum/species/mob_species = all_species[species]
```
Comment (skull132) [inline code/modules/client/preferences.dm]: Are two procs necessary? Find a way to remove one. If preferences.getMinAge() is used somewhere, then keep it. The lower class items can access it with pref.getMinAge().

File: code/modules/mob/living/carbon/human/examine.dm [hunk 5]
```diff
@@ -38,10 +38,17 @@
...
 	msg += "<EM>[src.name]</EM>"
+
 	if(species.name != "Human")
 		msg += ", a <b><font color='[species.flesh_color]'>[species.name]</font></b>"
 	msg += "!\n"
 
+	if (species && species.has_organ["ipc tag"] && internal_organs_by_name["ipc tag"])
+		msg += "[T.He] [T.is] wearing a tag designating them as Integrated Positronic Chassis <b>[src.real_name]</b>.\n"
+
+	if (species && species.has_organ["ipc tag"] && !internal_organs_by_name["ipc tag"])
```
Comment (skull132) [inline code/modules/mob/living/carbon/human/examine.dm]: Not sure if necessary. Finding the dudes in question should be a hunt, no?

File: code/modules/mob/living/carbon/human/species/station/machine_subspecies.dm [hunk 6]
```diff
@@ -0,0 +1,228 @@
...
+			H.m_intent = "walk"
+			H.hud_used.move_intent.update_move_icon(H)
+			H << span("danger", "ERROR: Power reserves depleted, emergency shutdown engaged. Backup power will come online in 60 seconds, initiate charging as primary directive.")
+			playsound(H.loc, 'sound/machines/buzz-two.ogg', 100, 0)
+	return 0
+
+/datum/species/machine/shell/handle_death(var/mob/living/carbon/human/H)
+	..()
+
+/datum/species/machine/shell/get_bodytype()
```
Comment (skull132) [inline code/modules/mob/living/carbon/human/species/station/machine_subspecies.dm]: This (56 - 57) can be deleted. It's inherited directly from the parent.

File: code/modules/mob/living/carbon/human/species/station/machine_subspecies.dm [hunk 7]
```diff
@@ -0,0 +1,228 @@
...
+			return 1
+		else
+			H.Weaken(30)
+			H.m_intent = "walk"
+			H.hud_used.move_intent.update_move_icon(H)
+			H << span("danger", "ERROR: Power reserves depleted, emergency shutdown engaged. Backup power will come online in 60 seconds, initiate charging as primary directive.")
+			playsound(H.loc, 'sound/machines/buzz-two.ogg', 100, 0)
+	return 0
+
+/datum/species/machine/shell/handle_death(var/mob/living/carbon/human/H)
```
Comment (skull132) [inline code/modules/mob/living/carbon/human/species/station/machine_subspecies.dm]: If a proc simply calls `..()` without additional actions, you can delete its definition.

File: code/modules/mob/living/carbon/human/species/station/machine_subspecies.dm [hunk 8]
```diff
@@ -0,0 +1,228 @@
...
+
+	heat_discomfort_level = 400
+	heat_discomfort_strings = list(
+		"Your CPU temperature probes warn you that you are approaching critical heat levels!",
+		"Your synthetic flesh crawls in the heat, swelling into a disgusting morass of plastic."
+		)
+
+	appearance_flags = HAS_HAIR_COLOR | HAS_SKIN_TONE | HAS_EYE_COLOR |  HAS_SKIN_COLOR | HAS_FBP
+
+/datum/species/machine/shell/handle_post_spawn(var/mob/living/carbon/human/H)
```
Comment (skull132) [inline code/modules/mob/living/carbon/human/species/station/machine_subspecies.dm]: delete

File: code/modules/mob/living/carbon/human/species/station/machine_subspecies.dm [hunk 9]
```diff
@@ -0,0 +1,228 @@
...
+			return 1
+		else
+			H.Weaken(60)
+			H.m_intent = "walk"
+			H.hud_used.move_intent.update_move_icon(H)
+			H << span("danger", "ERROR: Power reserves depleted, emergency shutdown engaged. Backup power will come online in two minutes, initiate charging as primary directive.")
+			playsound(H.loc, 'sound/machines/buzz-two.ogg', 100, 0)
+	return 0
+
+/datum/species/machine/industrial/handle_death(var/mob/living/carbon/human/H)
```
Comment (skull132) [inline code/modules/mob/living/carbon/human/species/station/machine_subspecies.dm]: delete

File: code/modules/mob/living/carbon/human/species/station/machine_subspecies.dm [hunk 10]
```diff
@@ -0,0 +1,228 @@
...
+			return 1
+		else
+			H.Weaken(5)
+			H.m_intent = "walk"
+			H.hud_used.move_intent.update_move_icon(H)
+			H << span("danger", "ERROR: Power reserves depleted, emergency shutdown engaged. Backup power will come online in 10 seconds, initiate charging as primary directive.")
+			playsound(H.loc, 'sound/machines/buzz-two.ogg', 100, 0)
+	return 0
+
+/datum/species/machine/handle_death(var/mob/living/carbon/human/H)
```
Comment (skull132) [inline code/modules/mob/living/carbon/human/species/station/machine_subspecies.dm]: This is good.

File: code/modules/mob/new_player/new_player.dm [hunk 11]
```diff
@@ -420,6 +420,10 @@
 			if(is_species_whitelisted(chosen_species) || has_admin_rights())
 				new_character = new(loc, use_species_name)
 
+				if (istype(chosen_species, /datum/species/machine))
```
Comment (skull132) [inline code/modules/mob/new_player/new_player.dm]: At this point, would it be worth bringing it to `datum/species/machine/handle_post_spawn(var/mob/living/carbon/human/H)`?

```
datum/species/machine/handle_post_spawn(var/mob/living/carbon/human/H)
	H.gender = NEUTER
	. = ..()
	check_tag(H, H.client)
```

File: code/modules/mob/new_player/preferences_setup.dm [hunk 12]
```diff
@@ -196,8 +196,17 @@ datum/preferences
...
-		var/g = "m"
-		if(gender == FEMALE)	g = "f"
+		var/genderhead = "head"
+		var/gendergroin = "groin"
+		var/genderbust = "torso"
+		if(gender == FEMALE)
+			genderhead = "head_f"
+			gendergroin = "groin_f"
+			genderbust = "torso_f"
+		if(gender == MALE)
```
Comment (skull132) [inline code/modules/mob/new_player/preferences_setup.dm]: Else if please.

File: code/modules/projectiles/gun.dm [hunk 13]
```diff
@@ -1,4 +1,4 @@
-/*
+h/*
```
Comment (skull132) [inline code/modules/projectiles/gun.dm]: Why isn't this throwing a compile error? Delete the h.

File: code/modules/client/preference_setup/general/01_basic.dm (hunk 1 above)
Comment (LordFowl) [inline code/modules/client/preference_setup/general/01_basic.dm]: Great idea.

File: code/modules/client/preference_setup/general/03_body.dm (hunk 2 above)
Comment (LordFowl) [inline code/modules/client/preference_setup/general/03_body.dm]: Why not? I've excluded no species. Let the loremin have feedback, but I don't really see any particular reason to exclude it. Shells are noticeably poor mimics anyways.

File: code/modules/client/preference_setup/general/03_body.dm (hunk 3 above)
Comment (LordFowl) [inline code/modules/client/preference_setup/general/03_body.dm]: What?

File: code/modules/client/preferences.dm (hunk 4 above)
Comment (LordFowl) [inline code/modules/client/preferences.dm]: Great idea.

File: code/modules/mob/living/carbon/human/examine.dm (hunk 5 above)
Comment (LordFowl) [inline code/modules/mob/living/carbon/human/examine.dm]: I thought it added a little bit of flavour. Can remove, though.

File: code/modules/mob/new_player/new_player.dm (hunk 11 above)
Comment (LordFowl) [inline code/modules/mob/new_player/new_player.dm]: Sure! Makes it more standardized.

File: code/modules/mob/living/carbon/human/species/station/station.dm [hunk 14]
```diff
@@ -430,11 +441,64 @@
...
+		if (query.NextRow())
+			status = text2num(query.item[1])
+		else
+			var/DBQuery/log_query = dbcon.NewQuery("INSERT INTO ss13_ipc_tracking (player_ckey, character_name) VALUES (:ckey, :character_name)")
+			log_query.Execute(query_details)
+
+		if (!status)
+			new_machine.internal_organs_by_name.Remove("ipc tag")
+			new_machine.internal_organs.Remove(tag)
+			del(tag)
```
Comment (skull132) [inline code/modules/mob/living/carbon/human/species/station/station.dm]: del(). Needs to be qdel().

File: code/modules/mob/new_player/new_player.dm (hunk 11 above)
Comment (skull132) [inline code/modules/mob/new_player/new_player.dm]: I think you forgot to remove this bit.

**Pull request merged**